presto-hive/src/main/java/com/facebook/presto/hive/orc/OrcSelectivePageSourceFactory.java

```diff
@@ -376,55 +379,82 @@ public static OrcSelectivePageSource createOrcPageSource(
          * - columns used for filtering only are updated to prune subfields if filters don't use full column
          */
 
-        Map<Integer, Set<Subfield>> requiredSubfields = new HashMap<>();
+        Map<Integer, Set<Subfield>> outputSubfields = new HashMap<>();
         physicalColumns.stream()
                 .filter(column -> outputColumns.contains(column.getHiveColumnIndex()))
-                .forEach(column -> requiredSubfields.put(column.getHiveColumnIndex(), new HashSet<>(column.getRequiredSubfields())));
+                .forEach(column -> outputSubfields.put(column.getHiveColumnIndex(), new HashSet<>(column.getRequiredSubfields())));
 
-        for (int index : outputColumns) {
-            requiredSubfields.computeIfAbsent(index, v -> new HashSet<>());
-        }
-
-        ImmutableSet.Builder<Subfield> predicateSubfields = ImmutableSet.builder();
-        remainingPredicate.accept(new RequiredSubfieldsExtractor(new SubfieldExtractor(functionResolution, rowExpressionService.getExpressionOptimizer(), session)), predicateSubfields);
+        Map<Integer, Set<Subfield>> predicateSubfields = new HashMap<>();
+        SubfieldExtractor subfieldExtractor = new SubfieldExtractor(functionResolution, rowExpressionService.getExpressionOptimizer(), session);
+        remainingPredicate.accept(new RequiredSubfieldsExtractor(subfieldExtractor), subfield -> predicateSubfields.computeIfAbsent(columnIndices.get(subfield.getRootName()), v -> new HashSet<>()).add(subfield));
 
-        for (Subfield subfield : predicateSubfields.build()) {
-            int index = columnIndices.get(subfield.getRootName());
-            updateRequiredSubfields(requiredSubfields, index, subfield);
+        for (Map.Entry<Integer, Map<Subfield, TupleDomainFilter>> entry : tupleDomainFilters.entrySet()) {
+            predicateSubfields.computeIfAbsent(entry.getKey(), v -> new HashSet<>()).addAll(entry.getValue().keySet());
```

Comment:
nit: `HashSet::new`

Reply:
@highker James, `HashSet::new` doesn't work here, because it ends up being called with `entry.getKey()` as a parameter which is interpreted as initial capacity. I'm going to keep this as `v -> new HashSet()`.